dev/ring-buffers.lisp

```diff
@@ -9,48 +9,119 @@
+   (total-size :reader total-size
+               :initarg :total-size)))
+
+(defclass* ring-buffer-reverse (ring-buffer) ())
+
+(defun make-ring-buffer (size &optional last-in-first-out)
+  (make-instance (if last-in-first-out 'ring-buffer-reverse 'ring-buffer)
+                 :contents (make-array size)
+                 :total-size size))
 
 (defmethod make-container ((class (eql 'ring-buffer)) &rest args)
-   (let ((total-size (getf args :total-size 1)))
-     (remf args :total-size)
-     (make-ring-buffer total-size)))
+  (let ((total-size (getf args :total-size 1))
+        (last-in-first-out (getf args :last-in-first-out nil)))
+    (remf args :total-size)
+    (remf args :last-in-first-out)
+    (make-ring-buffer total-size last-in-first-out)))
 
 
 ;;?? the (first indexes) is odd...
 (defmethod item-at ((container ring-buffer) &rest indexes)
-   (declare (dynamic-extent indexes))
-   (svref (contents container)
-          (mod (first indexes) (total-size container))))
+  (declare (dynamic-extent indexes))
+  (let ((indexes (mapcar #'(lambda (index) (fifo-index container index))
+                         indexes)))
+    (svref (contents container)
+           (mod (first indexes) (total-size container)))))
+
+(defmethod item-at ((container ring-buffer-reverse) &rest indexes)
+  (declare (dynamic-extent indexes))
+  (let ((indexes (mapcar #'(lambda (index) (lifo-index container index))
+                         indexes)))
+    (svref (contents container)
+           (mod (first indexes) (total-size container)))))
 
-(defmethod item-at! ((container ring-buffer) value &rest indexes)
+(defmethod item-at! ((container ring-buffer-reverse) value &rest indexes)
   (declare (dynamic-extent indexes))
-  (setf (svref (contents container)
-                (mod (first indexes) (total-size container)))
-         value))
+  (let ((indexes (mapcar #'(lambda (index) (fifo-index container index))
+                         indexes)))
+    (setf (svref (contents container)
+                 (mod (first indexes) (total-size container)))
+          value)))
 
+(defmethod item-at! ((container ring-buffer-reverse) value &rest indexes)
+  (declare (dynamic-extent indexes))
+  (let ((indexes (mapcar #'(lambda (index) (lifo-index container index))
+                         indexes)))
+    (setf (svref (contents container)
+                 (mod (first indexes) (total-size container)))
+          value)))
+
+(defmethod lifo-index ((container ring-buffer-reverse) index)
+  "Return index converted to internal LIFO index, where items are ordered from newest to oldest."
+  (mod (1- (+ (buffer-start container)
+              (- (buffer-end container) (buffer-start container) index)))
+       (total-size container)))
+
+(defmethod fifo-index ((container ring-buffer) index)
+  "Return index converted to internal FIFO index, where items are ordered from oldest to newest."
+  (mod (+ (buffer-start container) index)
+       (total-size container)))
```

Comment:
Another suggestion: rename both `lifo-index` and `fifo-index` to the same name (e.g. `index` if that's available).  Then you can remove the `item-at*` ring-buffer-reverse specialization.  Makes sense?

Reply:
Yes... I should've done this from the jump. Oops.